lib/internal/modules/esm/resolve.js

```diff
@@ -1,26 +1,588 @@
+      break;
+    const packageConfig = getPackageConfig(fileURLToPath(packageJSONUrl), base);
+    if (packageConfig.exists) return packageConfig;
+
+    const lastPackageJSONUrl = packageJSONUrl;
+    packageJSONUrl = new URL('../package.json', packageJSONUrl);
+
+    // Terminates at root where ../package.json equals ../../package.json
+    // (can't just check "/package.json" for Windows support).
+    if (packageJSONUrl.pathname === lastPackageJSONUrl.pathname) break;
+  }
+  const packageConfig = {
+    exists: false,
+    main: undefined,
+    name: undefined,
+    isValid: true,
+    type: 'none',
+    exports: undefined
+  };
+  packageJSONCache.set(fileURLToPath(packageJSONUrl), packageConfig);
+  return packageConfig;
+}
+
+/*
+ * Legacy CommonJS main resolution:
+ * 1. let M = pkg_url + (json main field)
+ * 2. TRY(M, M.js, M.json, M.node)
+ * 3. TRY(M/index.js, M/index.json, M/index.node)
+ * 4. TRY(pkg_url/index.js, pkg_url/index.json, pkg_url/index.node)
+ * 5. NOT_FOUND
+ */
+function fileExists(url) {
+  return tryStatSync(fileURLToPath(url)).isFile();
+}
+
+function legacyMainResolve(packageJSONUrl, packageConfig) {
+  let guess;
+  if (packageConfig.main !== undefined) {
+    // Note: fs check redundances will be handled by Descriptor cache here.
+    if (fileExists(guess = new URL(`./${packageConfig.main}`,
+                                   packageJSONUrl))) {
+      return guess;
+    }
+    if (fileExists(guess = new URL(`./${packageConfig.main}.js`,
+                                   packageJSONUrl))) {
+      return guess;
+    }
+    if (fileExists(guess = new URL(`./${packageConfig.main}.json`,
+                                   packageJSONUrl))) {
+      return guess;
+    }
+    if (fileExists(guess = new URL(`./${packageConfig.main}.node`,
+                                   packageJSONUrl))) {
+      return guess;
+    }
+    if (fileExists(guess = new URL(`./${packageConfig.main}/index.js`,
+                                   packageJSONUrl))) {
+      return guess;
+    }
+    if (fileExists(guess = new URL(`./${packageConfig.main}/index.json`,
+                                   packageJSONUrl))) {
+      return guess;
+    }
+    if (fileExists(guess = new URL(`./${packageConfig.main}/index.node`,
+                                   packageJSONUrl))) {
+      return guess;
+    }
+    // Fallthrough.
+  }
+  if (fileExists(guess = new URL('./index.js', packageJSONUrl))) {
+    return guess;
+  }
+  // So fs.
+  if (fileExists(guess = new URL('./index.json', packageJSONUrl))) {
+    return guess;
+  }
+  if (fileExists(guess = new URL('./index.node', packageJSONUrl))) {
+    return guess;
+  }
+  // Not found.
+  return undefined;
+}
+
+function resolveExtensionsWithTryExactName(search) {
+  if (fileExists(search)) return search;
+  return resolveExtensions(search);
+}
+
+const extensions = ['.js', '.json', '.node', '.mjs'];
+function resolveExtensions(search) {
+  for (const extension of extensions) {
```

Comment:
using for..of will break if Array.prototype[Symbol.iterator] is deleted - this should probably be a map + find instead?

Reply:
```suggestion
  for (let i = 0; i < extensions.length; i++) {
    const extension = extensions[i];
```